### ISSUE TITLE
Add GUI RPC to reset project debts

### ISSUE THREAD

**Issue by romw** [BOINC/boinc-ng#131]
**Reported by Nicolas on 7 Apr 37320047 15:33 UTC**
Feature request: Add a GUI RPC to the client to reset project debts. Preferably to set them to an arbitrary value, not only to zero them.

From what I know about debts, they are adjusted so that the sum of them is 0. This adjustment be done immediately after this RPC. However, changing from a positive debt on a single project to 0 would make it go positive again (and bring other projects ''closer'' to zero). For this reason, my suggested RPC can set multiple projects at a time, not needing one RPC per project.

Suggested RPC:
```
<set_debt>
    <project>
        <project_url>http://project1</project_url>
        <long_term_debt>0</long_term_debt>
        <short_term_debt>0</short_term_debt>
    </project>
    <project>
        <project_url>http://project2</project_url>
        <long_term_debt>0</long_term_debt>
        <short_term_debt>0</short_term_debt>
    </project>
</set_debt>
```

Answer to that would be `<success/>` like from most "set" RPCs. Note that the client should handle a missing LTD or STD tag (so that the RPC can be used to only change one of them).

I noticed the output of `<get_project_status>` and `<get_state>` RPCs already have the debts; so we have a mechanism to "get" them already.

**Comment by romw** [comment]
**Commented by anonymous on 29 Jul 37321236 05:20 UTC**
Replying to [Nicolas](ticket:131):
> Feature request: Add a GUI RPC to the client to reset project debts. Preferably to set them to an arbitrary value, not only to zero them.
> 
> From what I know about debts, they are adjusted so that the sum of them is 0. This adjustment be done immediately after this RPC. However, changing from a positive debt on a single project to 0 would make it go positive again (and bring other projects ''closer'' to zero). For this reason, my suggested RPC can set multiple projects at a time, not needing one RPC per project.
> 
> Suggested RPC:
> ```
> <set_debt>
>     <project>
>         <project_url>http://project1</project_url>
>         <long_term_debt>0</long_term_debt>
>         <short_term_debt>0</short_term_debt>
>     </project>
>     <project>
>         <project_url>http://project2</project_url>
>         <long_term_debt>0</long_term_debt>
>         <short_term_debt>0</short_term_debt>
>     </project>
> </set_debt>
> ```
> 
> Answer to that would be `<success/>` like from most "set" RPCs. Note that the client should handle a missing LTD or STD tag (so that the RPC can be used to only change one of them).
> 
> I noticed the output of `<get_project_status>` and `<get_state>` RPCs already have the debts; so we have a mechanism to "get" them already.
A couple of the possible CPU scheduler modifications will need different things that need to be reset as a group.

**Comment by romw** [comment]
**Commented by boincadm on 13 Nov 37401467 12:53 UTC**
(In [12749]) - Add GUI RPC to set projects debts (short and long term).
    Lets you set all of them at once, as suggested by Nicolas Alvarez.
    fixes #131

client/
    gui_rpc_server_ops.C
html/user/
    sample_index.php
lib/
    boinc_cmd.C
    gui_rpc_client.h
    gui_rpc_client_ops.C

**Comment by romw** [comment]
**Commented by Nicolas on 4 Oct 37496871 02:13 UTC**
Following report on #302, I noticed the RPC isn't working. I tried this with [guirpc_debug](ClientMessages) turned on, and I saw the core client is sending an '''empty''' reply back. No `<success/>` but no `<error>` either.
```
2007-06-30 14:02:32 [[guirpc_debug](---]) GUI RPC Command = '<boinc_gui_rpc_request>
<set_debts>
    <project>
        <master_url>http://climateprediction.net/</master_url>
        <short_term_debt>0.000000</short_term_debt>
        <long_term_debt>0.000000</long_term_debt>
    </project>
</set_debts>
</boinc_gui_rpc_request>
^C'
2007-06-30 14:02:32 [[guirpc_debug](---]) GUI RPC reply: '<boinc_gui_rpc_reply>
</boinc_gui_rpc_reply>
^C'
```

What [boinc_cmd](BoincCmd) is sending seems to be correct. I looked at the way [set_debt](source:trunk/boinc/client/gui_rpc_server_ops.C@12850#L869) expects the XML, and I think it should work with that XML input (although I don't know about the details of the XML parser API). It's strange that it doesn't work but no `<error>` is returned either.

**Comment by romw** [comment]
**Commented by davea on 5 Mar 37497010 15:33 UTC**
(In [13065]) - GUI RPC: debug set_debts handler.  Fixes #131

client/
    gui_rpc_server_ops.C

**Comment by romw** [comment]
**Commented by Nicolas on 21 May 37590128 19:33 UTC**
BOINC client windows_intelx86 5.10.13, RPC seems to be still broken. I will try 5.10.17 later; I was sticking to .13 trying to track down a different problem (AMS-related crash).

**Comment by romw** [comment]
**Commented by Nicolas on 7 Dec 37637584 23:06 UTC**
I don't see any change on the client_state.xml after calling the set_debt RPC. BOINC 5.10.18, Windows.

**Comment by romw** [comment]
**Commented by Nicolas on 9 Oct 38104463 00:26 UTC**
Replying to [Nicolas](comment:7):
> I don't see any change on the client_state.xml after calling the set_debt RPC. BOINC 5.10.18, Windows.
I don't know when this was fixed, but it definitely works on 5.10.3x.

**Comment by romw** [comment]
**Commented by Nicolas on 13 Apr 38104469 21:46 UTC**
Tracked it down. It was indeed fixed in [but applied to the 5.10 branch in [13211](13065],), fix available since 5.10.14.
